### ISSUE TITLE
cnvkit.py target antitarget do not run in parallel

### ISSUE THREAD

**Issue by apastore** [bcbio/bcbio-nextgen#2280]
Hi I just realized that the Sample-target.bed and Sample-antitarget.bed in cnvkit do not run in parallel, but one sample at once in a ipython run.

Am I doing something wrong? or this is the standard behavior?

Thanks a lot!

```
[2018-02-20T18:26Z] ls11: export TMPDIR=./work/bcbiotx/tmpoMpC3w && /bcbio/anaconda/bin/cnvkit.py target ./work/bedprep/cleaned-MMRF_1223_4_BM_CD138pos_T1_KHS5U_L11073-variant_regions.quantized-vrsubset-callableblocks.bed --split -o ./work/bcbiotx/tmpoMpC3w/MMRF_1223_2_PB_Whole_C3_KBS5U_L05821-target.bed --avg-size 250
[2018-02-20T18:27Z] ls11: export TMPDIR=./work/bcbiotx/tmpwcysJW && /bcbio/anaconda/bin/cnvkit.py antitarget -g ./work/regions/MMRF_1223_1-analysis_blocks.bed ./work/bedprep/cleaned-MMRF_1223_4_BM_CD138pos_T1_KHS5U_L11073-variant_regions.quantized-vrsubset-callableblocks.bed -o ./work/bcbiotx/tmpwcysJW/MMRF_1223_2_PB_Whole_C3_KBS5U_L05821-antitarget.bed --avg-size 1000000
[2018-02-20T18:27Z] ls11: export TMPDIR=./work/bcbiotx/tmp5oHetH && /bcbio/anaconda/bin/cnvkit.py target ./work/bedprep/cleaned-MMRF_1817_1_PB_Whole_C3_KBS5U_L05568-variant_regions.quantized-vrsubset-callableblocks.bed --split -o ./work/bcbiotx/tmp5oHetH/MMRF_1817_1_PB_Whole_C3_KBS5U_L05568-target.bed --avg-size 250
[2018-02-20T18:28Z] ls11: export TMPDIR=./work/bcbiotx/tmpwv5mlh && /bcbio/anaconda/bin/cnvkit.py antitarget -g ./work/regions/MMRF_1817-analysis_blocks.bed ./work/bedprep/cleaned-MMRF_1817_1_PB_Whole_C3_KBS5U_L05568-variant_regions.quantized-vrsubset-callableblocks.bed -o ./work/bcbiotx/tmpwv5mlh/MMRF_1817_1_PB_Whole_C3_KBS5U_L05568-antitarget.bed --avg-size 1000000
[2018-02-20T18:29Z] ls11: export TMPDIR=./work/bcbiotx/tmpvWAEVw && /bcbio/anaconda/bin/cnvkit.py target ./work/bedprep/cleaned-MMRF_1817_1_PB_Whole_C3_KBS5U_L05568-variant_regions.quantized-vrsubset-callableblocks.bed --split -o ./work/bcbiotx/tmpvWAEVw/MMRF_1817_1_BM_CD138pos_T1_KBS5U_L05541-target.bed --avg-size 250
[2018-02-20T18:30Z] ls11: export TMPDIR=./work/bcbiotx/tmpYFXGtp && /bcbio/anaconda/bin/cnvkit.py antitarget -g ./work/regions/MMRF_1817-analysis_blocks.bed ./work/bedprep/cleaned-MMRF_1817_1_PB_Whole_C3_KBS5U_L05568-variant_regions.quantized-vrsubset-callableblocks.bed -o ./work/bcbiotx/tmpYFXGtp/MMRF_1817_1_BM_CD138pos_T1_KBS5U_L05541-antitarget.bed --avg-size 1000000 
```

**Comment by chapmanb** [comment]
Thanks for the question. These steps are not parallelized across multiple machines so that the binning process uses consistent sizes across batched samples. The steps are pretty quick; it looks like in your case they take a minute or less each. As a result we haven't though to complicate things by having separate size estimation and parallelization but leave it all happening in one process. Please let us know if you're running into any bottlenecks and happy to reevaluate.

**Comment by apastore** [comment]
Thanks for your reply. I have a quite large batch (2000 sample). My first ideas was to split Tumor normal pair in set of 100 pairs.  Since cnvkit now use a background generated from all samples I decided to start a run with all samples together.... so the 1 minute is summing up ;) (and I did not see that coming). 

May I ask your opinion on this? 

Thanks!

**Comment by chapmanb** [comment]
Thanks for following up with more details about your setup. Practically, I'd suggest using the smaller batches rather than trying to put these all together. A hundred sample pairs together would give you a reasonable background for the samples. We haven't done a ton of testing with larger batch sizes like that and I think you'll run into additional bottlenecks both with ipython scaling and CNVkit steps. We're working on enabling these larger runs with the move to CWL and hopefully will have more and better scaling in the future.

**Comment by apastore** [comment]
Thanks a lot! I have change the setup as you suggested.

**Comment by apastore** [comment]
Hi Brad, I do not understand exactly how the ipython cluster is configured, but I am just wondering if would not be possible to downsize the number of engine node during the cnvkit input preparation. I just see all these Idle CPUs...
I definitely understand that this is not a priority and the CWL workflow will sort these things out. 
Thanks a lot!


with profile.report("callable regions", dirs):
            samples = run_parallel("prep_samples", [samples])
            samples = run_parallel("postprocess_alignment", samples)
            samples = run_parallel("combine_sample_regions", [samples])
            samples = run_parallel("calculate_sv_bins", [samples])
            samples = run_parallel("calculate_sv_coverage", samples)
            samples = run_parallel("normalize_sv_coverage", [samples])
            samples = region.clean_sample_data(samples)

**Comment by chapmanb** [comment]
Thanks for the suggestion, this is something we tackle better in CWL since each step is it's own submission to the cluster and will use exactly the resources it needs. In the current implementation we don't have the ability to resize the cluster on demand for different steps that require everything bundled together since we use a static IPython setup for the full process. This has pluses and minuses: the advantage is that you can keep the allocated cores you have on a busy cluster for the remaining steps, the downside is that they're idle during steps that do not distribute out over the samples. Sorry this is not optimal for your needs now but hope this explains the design reasoning and future improvements.

**Comment by apastore** [comment]
Sure thanks!

**Comment by roryk** [comment]
Thanks, Brad's answer still holds, so closing.
